### PR TITLE
Improve jenkins pipeline stability

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
3 changes:
1. Ctest re-runs failed tests
2. Improves the nightly jenkinsfile pipeline that it updates the conda-recipes repository with the correct git sha and corresponding version number for conda-build, before running a conda-build.
3. Adds an osx-ci CMakePreset that ensures osx runs in release mode so it doesn't timeout on unittests.

**To test:**
Code-review

*There is no associated issue.*

*This does not require release notes* because **Developer Facing**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: buildconfig/Jenkins/Conda/update-conda-recipes.sh [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+sed -i '/{% set version =/c\{% set version = "'$VERSION'" %}' ./meta.yaml
+sed -i '/  sha256: /c\  sha256: '$SHA256'' ./meta.yaml
+cd ../mantidworkbench
+sed -i '/{% set git_commit =/c\{% set git_commit = "'$LATEST_GIT_SHA'" %}' ./meta.yaml
+sed -i '/{% set version =/c\{% set version = "'$VERSION'" %}' ./meta.yaml
+sed -i '/  sha256: /c\  sha256: '$SHA256'' ./meta.yaml
+
+cd ../..
+git config user.name ${GITHUB_USER_NAME}
+git config user.email ${GITHUB_USER_NAME}@hosting.com
```
Comment (DavidFair) [inline buildconfig/Jenkins/Conda/update-conda-recipes.sh:71]: Is `@hosting.com` correct?

File: buildconfig/Jenkins/Conda/update-conda-recipes.sh [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    exit 1
+fi
+SHA_RESULT=$(command openssl sha256 $SOURCE_FILE)
+STRING_ARRAY=($SHA_RESULT)
+SHA256=${STRING_ARRAY[1]}
+echo $SHA256
+rm -rf $SOURCE_FILE
+
+# Clone conda-recipes
+git clone https://${GITHUB_USER_NAME}:${GITHUB_ACCESS_TOKEN}@github.com/mantidproject/conda-recipes.git
```
Comment (DavidFair) [inline buildconfig/Jenkins/Conda/update-conda-recipes.sh:54]: Do we sanitize the access token from the logs?

Comment (Pasarus) [reply buildconfig/Jenkins/Conda/update-conda-recipes.sh:54]: It auto sanitizes I think, this is given a discussion from @martyngigg

File: buildconfig/Jenkins/Conda/update-conda-recipes.sh (hunk 1 above)
Comment (Pasarus) [reply buildconfig/Jenkins/Conda/update-conda-recipes.sh:71]: It's not, I will replace it with @mantidproject.org, @hosting.org was likely a placeholder I left in by mistake.

File: buildconfig/Jenkins/Conda/nightly_build_and_deploy.jenkinsfile [hunk 3]
```diff
@@ -71,6 +75,12 @@ pipeline {
                 }
             }
         }
+        stage('Update conda-recipes repository') {
+            agent { label 'master' }
```
Comment (martyngigg) [inline buildconfig/Jenkins/Conda/nightly_build_and_deploy.jenkinsfile:79]: It'd be great to reduce the work the central node is required to do as it's a potential security issue. Now we have tht credential access in the job can we remove the restriction?

Comment (Pasarus) [reply buildconfig/Jenkins/Conda/nightly_build_and_deploy.jenkinsfile:79]: Uhhh yes we can, I will change it to any linux node so we don't clog OSX.

File: buildconfig/Jenkins/Conda/update-conda-recipes.sh (hunk 2 above)
Comment (martyngigg) [reply buildconfig/Jenkins/Conda/update-conda-recipes.sh:54]: Jenkins stars everything if it's injected as a password or a credential when it's referenced. See a run of [open-release-testing](https://builds.mantidproject.org/job/open-release-testing) for example.

File: buildconfig/Jenkins/Conda/update-conda-recipes.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+cd conda-recipes/recipes/mantid
+sed -i '/{% set git_commit =/c\{% set git_commit = "'$LATEST_GIT_SHA'" %}' ./meta.yaml
```
Comment (martyngigg) [inline buildconfig/Jenkins/Conda/update-conda-recipes.sh:57]: Can we capture these 3 lines in a local bash function to remove the duplication?

Comment (Pasarus) [reply buildconfig/Jenkins/Conda/update-conda-recipes.sh:57]: Yes, I will do that now.

**Review by martyngigg — APPROVED** [review]
I'm happy with this.

**Pull request merged**